File: src/main/java/com/zepben/testutils/junit/SystemLogExtension.java
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package com.zepben.testutils.junit;

import com.zepben.annotations.EverythingIsNonnullByDefault;
import org.junit.jupiter.api.extension.AfterEachCallback;
import org.junit.jupiter.api.extension.BeforeEachCallback;
import org.junit.jupiter.api.extension.ExtensionContext;

import java.io.*;
import java.nio.charset.Charset;
import java.util.function.Consumer;

import static java.lang.System.lineSeparator;

/**
 * JUnit5 Extension implementation of https://github.com/stefanbirkner/system-rules log handlers
 * <p>
 * See tests for example usage.
 */
@EverythingIsNonnullByDefault
@SuppressWarnings("UnusedReturnValue")
public class SystemLogExtension implements BeforeEachCallback, AfterEachCallback {

    public static SystemLogExtension SYSTEM_OUT = new SystemLogExtension(System.out, System::setOut);
    public static SystemLogExtension SYSTEM_ERR = new SystemLogExtension(System.err, System::setErr);

    private final PrintStream originalStream;
    private final Consumer<PrintStream> streamReplacer;
    private final LogWrapper logWrapper;

    public SystemLogExtension clearCapturedLog() {
        logWrapper.captureLog.reset();
        return this;
    }

    public SystemLogExtension captureLog() {
        logWrapper.captureLogEnabled = true;
        return this;
    }

    public SystemLogExtension stopCapturingLog() {
        logWrapper.captureLogEnabled = false;
        return this;
    }

    public SystemLogExtension mute() {
        logWrapper.originalStreamEnabled = false;
        logWrapper.failureLogEnabled = false;
        return this;
    }

    public SystemLogExtension unmute() {
        logWrapper.originalStreamEnabled = true;
        logWrapper.failureLogEnabled = false;
        return this;
    }

    public SystemLogExtension muteOnSuccess() {
        mute();
        logWrapper.failureLogEnabled = true;
        return this;
    }

    public String getLog() {
        try {
            return logWrapper.captureLog.toString(Charset.defaultCharset().name());
        } catch (UnsupportedEncodingException e) {
            throw new RuntimeException(e);
        }
    }

    public String[] getLogLines() {
        String log = getLog();
        if (log.equals(""))
            return new String[0];
        else
            return log.split(lineSeparator(), 0);
    }

    @Override
    public void beforeEach(ExtensionContext extensionContext) throws Exception {
        streamReplacer.accept(new PrintStream(logWrapper, true, Charset.defaultCharset().name()));
    }

    @Override
    public void afterEach(ExtensionContext extensionContext) throws Exception {
        streamReplacer.accept(originalStream);

        if (extensionContext.getExecutionException().isPresent())
            logWrapper.failureLog.writeTo(originalStream);

        logWrapper.failureLog.reset();
        logWrapper.captureLog.reset();
    }

    private SystemLogExtension(PrintStream originalStream, Consumer<PrintStream> streamReplacer) {
        this.originalStream = originalStream;
        this.streamReplacer = streamReplacer;
        this.logWrapper = new LogWrapper(originalStream);
    }

    private static class LogWrapper extends OutputStream {

        final OutputStream originalStream;
        final ByteArrayOutputStream failureLog = new ByteArrayOutputStream();
        final ByteArrayOutputStream captureLog = new ByteArrayOutputStream();
        boolean originalStreamEnabled = true;
        boolean failureLogEnabled = false;
        boolean captureLogEnabled = false;

        LogWrapper(OutputStream originalStream) {
            this.originalStream = originalStream;
        }

        @Override
        public void write(int b) throws IOException {
            if (originalStreamEnabled)
                originalStream.write(b);
            if (failureLogEnabled)
                failureLog.write(b);
            if (captureLogEnabled)
                captureLog.write(b);
        }

        @Override
        public void flush() throws IOException {
            originalStream.flush();
            //ByteArrayOutputStreams don't have to be closed
        }

        @Override
        public void close() throws IOException {
            originalStream.close();
            //ByteArrayOutputStreams don't have to be closed
        }

    }

}

File: src/main/kotlin/com/zepben/testutils/junit/SystemLogExtension.kt
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package com.zepben.testutils.junit

import org.junit.jupiter.api.extension.AfterAllCallback
import org.junit.jupiter.api.extension.AfterEachCallback
import org.junit.jupiter.api.extension.BeforeAllCallback
import org.junit.jupiter.api.extension.BeforeEachCallback
import org.junit.jupiter.api.extension.ExtensionContext
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.OutputStream
import java.io.PrintStream
import java.nio.charset.Charset
import java.util.function.Consumer

/**
 * JUnit5 Extension implementation of [https://github.com/stefanbirkner/system-rules](https://github.com/stefanbirkner/system-rules) log handlers
 *
 *
 * See tests for example usage.
 */
@Suppress("MemberVisibilityCanBePrivate")
class SystemLogExtension private constructor(
    private val originalStream: PrintStream,
    private val streamReplacer: Consumer<PrintStream>
) : BeforeAllCallback, BeforeEachCallback, AfterEachCallback, AfterAllCallback {

    val log: String
        get() = logWrapper.captureLog.toString(Charset.defaultCharset())

    val logLines: Array<String>
        get() =
            log.split(System.lineSeparator()).dropLastWhile { it.isBlank() }.toTypedArray()

    private val logWrapper: LogWrapper = LogWrapper(originalStream)
    private lateinit var initialSettings: LogWrapperSettings

    init {
        streamReplacer.accept(PrintStream(logWrapper, true, Charset.defaultCharset()))
    }

    fun clearCapturedLog(): SystemLogExtension {
        logWrapper.captureLog.reset()
        return this
    }

    fun captureLog(): SystemLogExtension {
        logWrapper.settings = logWrapper.settings.copy(captureLogEnabled = true)
        return this
    }

    fun stopCapturingLog(): SystemLogExtension {
        logWrapper.settings = logWrapper.settings.copy(captureLogEnabled = false)
        return this
    }

    fun mute(): SystemLogExtension {
        logWrapper.settings = logWrapper.settings.copy(originalStreamEnabled = false, failureLogEnabled = false)
        return this
    }

    fun unmute(): SystemLogExtension {
        logWrapper.settings = logWrapper.settings.copy(originalStreamEnabled = true, failureLogEnabled = false)
        return this
    }

    fun muteOnSuccess(): SystemLogExtension {
        mute()
        logWrapper.settings = logWrapper.settings.copy(failureLogEnabled = true)
        return this
    }

    override fun beforeAll(extensionContext: ExtensionContext) {
        streamReplacer.accept(PrintStream(logWrapper, true, Charset.defaultCharset()))
        initialSettings = logWrapper.settings
    }

    @Throws(Exception::class)
    override fun beforeEach(extensionContext: ExtensionContext) {
        logWrapper.settings = initialSettings
    }

    @Throws(Exception::class)
    override fun afterEach(extensionContext: ExtensionContext) {
        if (extensionContext.executionException.isPresent) logWrapper.failureLog.writeTo(originalStream)

        logWrapper.settings = initialSettings

        logWrapper.failureLog.reset()
        logWrapper.captureLog.reset()
    }

    override fun afterAll(extensionContext: ExtensionContext) {
        logWrapper.settings = initialSettings
        streamReplacer.accept(originalStream)
    }

    private class LogWrapper(
        val originalStream: OutputStream,
    ) : OutputStream() {

        val failureLog = ByteArrayOutputStream()
        val captureLog = ByteArrayOutputStream()

        var settings = LogWrapperSettings(
            originalStreamEnabled = true,
            captureLogEnabled = false,
            failureLogEnabled = false
        )

        @Throws(IOException::class)
        override fun write(b: Int) {
            if (settings.originalStreamEnabled) originalStream.write(b)
            if (settings.failureLogEnabled) failureLog.write(b)
            if (settings.captureLogEnabled) captureLog.write(b)
        }

        @Throws(IOException::class)
        override fun flush() {
            originalStream.flush()
            //ByteArrayOutputStreams don't have to be closed
        }

        @Throws(IOException::class)
        override fun close() {
            originalStream.close()
            //ByteArrayOutputStreams don't have to be closed
        }

    }

    private data class LogWrapperSettings(
        val originalStreamEnabled: Boolean,
        val captureLogEnabled: Boolean,
        val failureLogEnabled: Boolean
    )

    companion object {

        @JvmField
        var SYSTEM_OUT = SystemLogExtension(System.out) { out: PrintStream? -> System.setOut(out) }

        @JvmField
        var SYSTEM_ERR = SystemLogExtension(System.err) { err: PrintStream? -> System.setErr(err) }

    }

}
